### DISCUSSION TITLE
GUI crashes after fresh pip install in mamba env

### DISCUSSION THREAD

**Discussion by antortjim** [talmolab/sleap#1469] [Help!]
Dear SLEAP developers

I am unable to run the SLEAP GUI after a fresh install using the instructions on the website due to (I believe) an issue with PySide2 and opencv. What can I do to fix it? Thank you!

## Reproduce
```
mamba create --name sleap pip python=3.7.12 cudatoolkit=11.3 cudnn=8.2
mamba activate sleap
pip install sleap==1.3.1
```

Then I run `sleap-label` and I get

```
Saving config: /home/vibflysleep/.sleap/1.3.1/preferences.yaml
QObject::moveToThread: Current thread (0x56480c306ca0) is not the object's thread (0x56480c0320f0).
Cannot move to target thread (0x56480c306ca0)

qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "/home/vibflysleep/mambaforge/envs/sleap/lib/python3.7/site-packages/cv2/qt/plugins" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: xcb, eglfs, linuxfb, minimal, minimalegl, offscreen, vnc, wayland-egl, wayland, wayland-xcomposite-egl, wayland-xcomposite-glx, webgl.
```

## Partial solution

Swapping `opencv-python` for `opencv-python-headless` allows me to at least launch the app

```
# https://stackoverflow.com/a/73457305
pip uninstall opencv-python opencv-python-headless opencv-contrib-python
pip install opencv-python-headless==4.5.5.64 # this is the openv-python version I had
```
```
sleap-label
# The app launches!
```
but it crashes whenever I enter a number in any numeric input like for example, the "Samples per video" box when generating Labeling Suggestions. I can increase/decrease it with the arrows, but I cannot directly enter a different number.

## System

```
python -c "import sleap; sleap.versions()"
SLEAP: 1.3.1
TensorFlow: 2.8.4
Numpy: 1.21.6
Python: 3.7.12
OS: Linux-5.19.0-50-generic-x86_64-with-debian-bookworm-sid
```

**Comment by antortjim** [comment]
I am also trying to change to different versions of PySide2 (within the range `PySide2>=5.13.2,<=5.14.1;`) but I keep getting weird errors when trying to modify the numeric input

```
/home/vibflysleep/mambaforge/envs/sleap3/bin/python3.7: symbol lookup error: /home/vibflysleep/mambaforge/envs/sleap3/lib/python3.7/site-packages/PySide2/Qt/plugins/platforminputcontexts/libibusplatforminputcontextplugin.so: undefined symbol: _ZdlPvm, version Qt_5
```

**Comment by roomrys** [comment]
Hi @antortjim,

Yes, I second that the issue lies with a `Qt` conflict between `PySide` and `opencv`. Using `opencv-headless` ensures that there is no `Qt` conflict. The third player here that puts us in a bit of a pickle is `imgaug` (another SLEAP dependency) which requires/uses `opencv` non-headless.

While I haven't run into the problem above on my local Linux, our Github Actions has run into this issue... I feel very negative that this is showing up locally as the workaround to continue development was to uninstall all versions of opencv __post SLEAP installation__ and install desired version:
https://github.com/talmolab/sleap/blob/fcc8c285186a8bde19996df37a510996eec29c68/.github/workflows/ci.yml#L99-L100
and also directly point to the Qt path plugin:
https://github.com/talmolab/sleap/blob/fcc8c285186a8bde19996df37a510996eec29c68/tests/gui/test_app.py#L10-L14

Very very hacky and not for production....

Since I haven't seen this in the GUI, I also can't comment on the numerical input, but I suppose 🤞I see it soon.

Thanks,
Liezl

## Related (affected development):
- #1454 ([action](https://github.com/talmolab/sleap/actions/runs/5895812564/job/15992499755?pr=1454))
- #1258
- #1460

**Comment by antortjim** [comment]
Hi @roomrys thanks for the quick reply

I am unable to get a fully working SLEAP GUI. The best I got is as I said to launching the app and having it crash whenever I modify an input directly. And this is only thanks to installing `opencv-contrib-python` and having the environment modification you specify in your snippets above. Without this, it will not even launch.

I made a quick video showing the error. The app crashes the moment I try to remove the last digit by pressing the backspace
[error.zip](https://github.com/talmolab/sleap/files/12437493/error.zip)

**Comment by antortjim** [comment]
OK I think this issue is solved with the following environment

```
mamba create --name sleap4 pip python=3.7.12 cudatoolkit=11.3 cudnn=8.2
mamba activate sleap4
pip install sleap==1.3.1


git clone git@github.com:talmolab/sleap@0ef52cd577fd4ef9331b745a62ff9a857d1dd61f
cd sleap

pip uninstall sleap
pip install -e .

mamba install pyside2=5.15.6 -c conda-forge
pip install  opencv-python==4.7.0.68
pip install opencv-python-headless==4.7.0.68
```

However, I haven't fully tested it, and I don't know if some ML operation is broken in this environment. But at least the GUI does not crash anymore.
None of the `PySide2` versions available from pip seem to work, only the one from `conda-forge`.
I have to install sleap first from pip and then from github because otherwise many dependencies are not installed (probably because the `requirements.txt` file in `0ef52cd577fd4ef9331b745a62ff9a857d1dd61f` does not list them anymore (`cattr`, `sklearn`, `psutil`, etc)

**Reply by roomrys** [reply]
> I have to install sleap first from pip and then from github because otherwise many dependencies are not installed (probably because the requirements.txt file in 0ef52cd577fd4ef9331b745a62ff9a857d1dd61f does not list them anymore (cattr, sklearn, psutil, etc)

I will also comment here that the pip package for 1.3.1 comes from 1a0404c0ffae7b248eb360562b0bb95a42a287b6. However, the conda build was failing, so we did a dependency deep-dive to get a conda package from 8c6bdb407c80848db7671eb0e404d0ca32a81b5e - opting to leave the pypi build as is and deal with that dependency hurdle in the 1.3.2 release (#1454 - the 1.3.2 must-have).
